### PR TITLE
API Filtering plus

### PR DESCRIPTION
**Pull request by abellotti**
Brandon's API Filtering PR https://github.com/ManageIQ/topological_inventory-api/pull/156 +
- quieting down hakiri - now using arel
- properly escaping SQL like patterns so we can also match characters that include %
- additional tests

### CODE REVIEW

File: .codeclimate.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+version: '2'
+checks:
+  method-count:
+    enabled: false
```
Comment (Fryguy) [inline .codeclimate.yml]: This should not be necessary...we likely just missed adding the general .codeclimate.yml to this.

Comment (Fryguy) [reply .codeclimate.yml]: You can ignore codeclimate for this PR as their issues are probably too strict for us.

File: lib/topological_inventory/api/error_document.rb [hunk 2]
```diff
@@ -1,19 +1,27 @@
 module TopologicalInventory
   module Api
-    class ErrorDocument < Hash
-      def initialize
-        self["errors"] = []
-      end
-
-      def add(status, message)
+    class ErrorDocument
```
Comment (Fryguy) [inline lib/topological_inventory/api/error_document.rb]: Let's leave this the way it was for now...too much refactoring in a PR adding a new feature and it's cluttering up the diff.

Comment (Fryguy) [reply lib/topological_inventory/api/error_document.rb]: Oh I see Brandon did this in his original commits...let's leave it then...I don't want to mess with his original commits.

File: .codeclimate.yml (hunk 1 above)
Comment (abellotti) [reply .codeclimate.yml]: ok, reverted this.

File: lib/topological_inventory/api/error_document.rb (hunk 2 above)
Comment (abellotti) [reply lib/topological_inventory/api/error_document.rb]: ok, we can always refactor  later.

File: app/controllers/application_controller/filter.rb [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      else
+        add_filter("eq", STRING_COMPARISON_KEYWORDS, k, val)
+      end
+    end
+
+    def add_filter(requested_comparator, allowed_comparators, key, value)
+      return unless attribute = attribute_for_key(key)
+      type = determine_string_attribute_type(attribute)
+
+      if requested_comparator.in?(["nil", "not_nil"])
```
Comment (Fryguy) [inline app/controllers/application_controller/filter.rb]: super minor, but we should extract these into a NIL_COMPARISON_KEYWORDS constants in a follow up PR.

**Comment by bdunne** [comment]
@abellotti Are there any tests for the SQL injection cases?

**Pull request merged**